### PR TITLE
Treat hidden code (#line hidden) as generated code for analyzers

### PR DESCRIPTION
**Pull request by mavasani**
**Customer scenario**

Users see analyzer diagnostics in hidden (non-user) code of .cshtml and .aspx pages for analyzers configured to skip generated code. This is considerable noise for users, especially given we don't offer light bulb/code fixes for them. We have heard customer reports and dogfooding feedback about hiding these diagnostics.

**Bugs this fixes:** 

Fixes https://github.com/dotnet/roslyn/issues/15903

**Workarounds, if any**

Users can suppress symbol diagnostics in hidden code using suppress message attributes. However, non-symbol diagnostics (e.g. those reported on using declarations) cannot be suppressed with suppress message attributes. So users would have to add pragma suppressions around hidden code.

**Risk**

We already have the APIs in place to configure and skip generated code analysis in the analyzer driver. This change just adds additional checks prior to analyzer callbacks to verify if the code being analyzed is hidden code location or not.

**Performance impact**

`<TBD>`: I will fill in after perf analysis. Likely neglible, as we cache upfront if tree has any hidden regions, and perform additional work only for files that have at least one hidden region.

**Is this a regression from a previous update?**

No.

**Root cause analysis:**

We missed hidden regions while working on generated code analysis feature. We are adding unit tests for the new functionality.

**How was the bug found?**

Customer reported.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
👍

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 1]
```diff
@@ -1298,6 +1304,38 @@ protected bool IsGeneratedCode(SyntaxTree tree)
...
+            {
+                return false;
+            }
+
+            Debug.Assert(_lazyTreesWithHiddenRegionsMap != null);
+
+            lock (_lazyTreesWithHiddenRegionsMap)
+            {
+                bool hasHiddenRegions;
+                if (!_lazyTreesWithHiddenRegionsMap.TryGetValue(tree, out hasHiddenRegions))
```
Comment (heejaechang) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: who put the value to the dictionary? looks like you forgot to add result to the map.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 2]
```diff
@@ -1298,6 +1304,38 @@ protected bool IsGeneratedCode(SyntaxTree tree)
...
         protected bool DoNotAnalyzeGeneratedCode => _doNotAnalyzeGeneratedCode;
 
+        // Location is in generated code if either the containing tree is a generated code file OR if it is a hidden source location.
+        protected bool IsGeneratedCodeLocation(Location location) =>
+            IsGeneratedCode(location.SourceTree) || IsHiddenSourceLocation(location);
+
+        protected bool IsHiddenSourceLocation(Location location) =>
+            location.IsInSource &&
+            HasHiddenRegions(location.SourceTree) &&
+            location.SourceTree.IsHiddenPosition(location.SourceSpan.Start);
```
Comment (heejaechang) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: why it doesn't check whole span? probably okay for venus case.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 3]
```diff
@@ -1298,6 +1304,38 @@ protected bool IsGeneratedCode(SyntaxTree tree)
...
+            }
+
+            Debug.Assert(_lazyTreesWithHiddenRegionsMap != null);
+
+            lock (_lazyTreesWithHiddenRegionsMap)
+            {
+                bool hasHiddenRegions;
+                if (!_lazyTreesWithHiddenRegionsMap.TryGetValue(tree, out hasHiddenRegions))
+                {
+                    hasHiddenRegions = tree.HasHiddenRegions();
```
Comment (heejaechang) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: just calling HasHiddenRegions seems require tree to create line directive map which will visit every single node/token/trivia in the tree. at least, this doesnt require text to be brought into memory.

File: src/Compilers/Core/Portable/Syntax/SyntaxTreeExtensions.cs [hunk 4]
```diff
@@ -94,5 +95,19 @@ private static int FindFirstDifference(string s1, string s2)
...
+        /// Returns <c>true</c> if the provided position is in a hidden region inaccessible to the user.
+        /// </summary>
+        public static bool IsHiddenPosition(this SyntaxTree tree, int position, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (!tree.HasHiddenRegions())
+            {
+                return false;
+            }
+
+            var lineVisibility = tree.GetLineVisibility(position, cancellationToken);
```
Comment (heejaechang) [inline src/Compilers/Core/Portable/Syntax/SyntaxTreeExtensions.cs]: GetLineVisibility seems require even SourceText to work. so it might be more expansive to do than we think?

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs (hunk 3 above)
Comment (heejaechang) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: since SyntaxNodeAction visit all nodes anyway, might not be an issue. but if it is, we could do simple text search for the pragma (ex, "#hidden") before calling HasHiddenRegions to make it a bit cheaper.

**Comment by heejaechang** [comment]
except forgetting to put result to _lazyTreesWithHiddenRegionsMap and perf test, LGTM

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs (hunk 1 above)
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: On second thought, I have decided to remove this map. The syntax tree already caches the line directive map, and the subsequent check for hidden entries should be cheap.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs (hunk 2 above)
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: There doesn't seem to be an API to check for the visibility of the span - also we probably shouldn't have diagnostics/declarations with locations across hidden and non-hidden regions.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs (hunk 3 above)
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: I put in a slight optimization in LineDirectiveMap to not realize the text unless there are any directives in the tree. I don't think we need text search as it would be language specific, plus majority of source files won't have line directives and hence are unaffected.

File: src/Compilers/Core/Portable/Syntax/SyntaxTreeExtensions.cs (hunk 4 above)
Comment (mavasani) [inline src/Compilers/Core/Portable/Syntax/SyntaxTreeExtensions.cs]: Yes, line directive map realizes the source text, but only if there are any line directives in tree, which should be rare.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: I think this is ok as well.

File: src/Compilers/Core/Portable/Syntax/LineDirectiveMap.cs [hunk 5]
```diff
@@ -34,11 +35,11 @@ protected LineDirectiveMap(SyntaxTree syntaxTree)
         {
             // Accumulate all the directives, in source code order
             var syntaxRoot = (SyntaxNodeOrToken)syntaxTree.GetRoot();
-            IEnumerable<TDirective> directives = syntaxRoot.GetDirectives<TDirective>(filter: ShouldAddDirective);
+            IList<TDirective> directives = syntaxRoot.GetDirectives<TDirective>(filter: ShouldAddDirective);
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Syntax/LineDirectiveMap.cs]: If we're touching this code, it would be much nicer for these to eb immutable arrays.  changing to this shouldn't be too hard.

Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Syntax/LineDirectiveMap.cs]: n/m.  i see how this works now.

File: src/Compilers/Core/Portable/Syntax/LineDirectiveMap.cs [hunk 6]
```diff
@@ -96,35 +97,37 @@ protected LineMappingEntry FindEntry(int lineNumber)
...
-            entries[index] = current;
+            entries.Add(current);
 
-            foreach (var directive in directives)
+            if (directives.Count > 0)
             {
-                current = GetEntry(directive, sourceText, current);
-                ++index;
-                entries[index] = current;
+                var sourceText = tree.GetText();
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Syntax/LineDirectiveMap.cs]: GetText can take a cancellatoin token.  Any chance we can thread cancellation along here?

**Comment by CyrusNajmabadi** [comment]
👍

File: src/Compilers/Core/Portable/Syntax/LineDirectiveMap.cs (hunk 6 above)
Comment (heejaechang) [inline src/Compilers/Core/Portable/Syntax/LineDirectiveMap.cs]: so, creating line directive map itself do require reading in text I guess.

**Comment by heejaechang** [comment]
I think it is good to have the map you removed. checking whether tree has hidden region might be cheap to called once, but not cheap to be called on every node/operation in the tree.

http://source.roslyn.io/#Microsoft.CodeAnalysis/Syntax/LineDirectiveMap.cs,83

**Comment by heejaechang** [comment]
also, now you need 2 sign off from compiler team.

**Comment by heejaechang** [comment]
@jaredpar @AlekseyTs @cston @vladima @agocke @dotnet/roslyn-compiler can you code review compiler side of changes? thank you

File: src/Compilers/Core/Portable/Syntax/LineDirectiveMap.cs (hunk 6 above)
Comment (mavasani) [inline src/Compilers/Core/Portable/Syntax/LineDirectiveMap.cs]: HasHiddenRegions is a public API, so I feel we should keep the current approach and add a public API with cancellation token if we see performance delays.

**Comment by mavasani** [comment]
@heejaechang I have added back the map. Thanks.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 7]
```diff
@@ -1298,6 +1304,39 @@ protected bool IsGeneratedCode(SyntaxTree tree)
 
         protected bool DoNotAnalyzeGeneratedCode => _doNotAnalyzeGeneratedCode;
 
+        // Location is in generated code if either the containing tree is a generated code file OR if it is a hidden source location.
+        protected bool IsGeneratedCodeLocation(Location location) =>
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: >IsGeneratedCodeLocation [](start = 23, length = 23)

IsGeneratedOrHiddenLocation?

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 8]
```diff
@@ -1298,6 +1304,39 @@ protected bool IsGeneratedCode(SyntaxTree tree)
...
+        protected bool IsHiddenSourceLocation(Location location) =>
+            location.IsInSource &&
+            HasHiddenRegions(location.SourceTree) &&
+            location.SourceTree.IsHiddenPosition(location.SourceSpan.Start);
+
+        private bool HasHiddenRegions(SyntaxTree tree)
+        {
+            Debug.Assert(tree != null);
+
+            if (_treatAllCodeAsNonGeneratedCode)
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: >_treatAllCodeAsNonGeneratedCode [](start = 16, length = 31)

Perhaps should check _lazyTreesWithHiddenRegionsMap instead and get rid of the assert below.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 9]
```diff
@@ -1298,6 +1304,39 @@ protected bool IsGeneratedCode(SyntaxTree tree)
...
+            {
+                return false;
+            }
+
+            Debug.Assert(_lazyTreesWithHiddenRegionsMap != null);
+
+            lock (_lazyTreesWithHiddenRegionsMap)
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: >lock (_lazyTreesWithHiddenRegionsMap) [](start = 12, length = 37)

Is there an advantage to not use concurrent dictionary here?

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs (hunk 7 above)
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: Sure, will change.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs (hunk 8 above)
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: Yes, sure

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs (hunk 9 above)
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: Not really, just followed the existing pattern.

File: src/Compilers/Core/Portable/Syntax/LineDirectiveMap.cs [hunk 10]
```diff
@@ -19,7 +20,7 @@ namespace Microsoft.CodeAnalysis
     internal abstract partial class LineDirectiveMap<TDirective>
         where TDirective : SyntaxNode
     {
-        protected readonly LineMappingEntry[] Entries;
+        protected readonly ImmutableArray<LineMappingEntry> Entries;
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Syntax/LineDirectiveMap.cs]: >ImmutableArray<LineMappingEntry> Entries; [](start = 27, length = 41)

This change creates too much churn to an escrow level change. Consider moving it to a separate PR targeting post RTM branch.

**Comment by AlekseyTs** [comment]
Done with review pass.

File: src/Compilers/Core/Portable/Syntax/LineDirectiveMap.cs (hunk 10 above)
Comment (mavasani) [inline src/Compilers/Core/Portable/Syntax/LineDirectiveMap.cs]: @AlekseyTs do you mean the change to this file or the entire PR? The change to this file just changes the types from array to ImmutableArray, and makes the fetching of source text more lazy.

Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Syntax/LineDirectiveMap.cs]: I suggest you to only keep the changes that affect behavior in the way you want it to change. Changing the type of this field doesn't appear to be related to any desired behavior change it can be done post RTM.

**Comment by mavasani** [comment]
Thanks @AlekseyTs, I have addressed your comments.

File: src/Compilers/Core/Portable/Syntax/LineDirectiveMap.cs [hunk 11]
```diff
@@ -102,18 +102,22 @@ protected int FindEntryIndex(int lineNumber)
 
         // Given the ordered list of all directives in the file, return the ordered line mapping
         // entry for the file. This always starts with the null mapped that maps line 0 to line 0.
-        private LineMappingEntry[] CreateEntryMap(SourceText sourceText, IEnumerable<TDirective> directives)
+        private LineMappingEntry[] CreateEntryMap(SyntaxTree tree, IList<TDirective> directives)
         {
             var entries = new LineMappingEntry[directives.Count() + 1];
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Syntax/LineDirectiveMap.cs]: >.Count() [](start = 57, length = 8)

Should we be using this method now?

File: src/Compilers/Core/Portable/Syntax/LineDirectiveMap.cs [hunk 12]
```diff
@@ -102,18 +102,22 @@ protected int FindEntryIndex(int lineNumber)
...
             var current = InitializeFirstEntry();
             var index = 0;
             entries[index] = current;
 
-            foreach (var directive in directives)
+            if (directives.Count > 0)
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Syntax/LineDirectiveMap.cs]: >directives.Count [](start = 16, length = 16)

It looks like we need this value twice. Should we cache and reuse it?

**Comment by AlekseyTs** [comment]
LGTM

File: src/Compilers/Core/Portable/Syntax/LineDirectiveMap.cs (hunk 12 above)
Comment (mavasani) [inline src/Compilers/Core/Portable/Syntax/LineDirectiveMap.cs]: Probably can be done as part of the cleanup to use ImmutableArray.

**Comment by mavasani** [comment]
Tagging @MattGertz for RTW approval.

**Comment by MattGertz** [comment]
It's not clear that this meets the RTW Escrow bar, which is focused on perf and crashes -- we have officially entered the "be paranoid" stage.  I will need a benefit vs. risk assessment (particularly number of users impacted and performance implications).

**Comment by mavasani** [comment]
@dotnet-bot test this please

**Comment by mavasani** [comment]
Rebased to post-dev15 branch.
